src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -333,27 +379,75 @@ void Connection::websocket_handshake_completion_handler(const std::string& proto
 
 void Connection::websocket_read_or_write_error_handler(std::error_code ec)
 {
+    logger.error("Read/Write error: %1", ec.message());
+    m_num_redirects = 0;
     read_or_write_error(ec); // Throws
 }
 
 
+bool Connection::update_location_parameters(const std::string_view& location)
+{
+    ProtocolEnvelope protocol = {};
+    std::string address;
+    port_type port = {};
+    std::string path;
+    if (!m_client.decompose_server_url(std::string(location), protocol, address, port, path)) { // Throws
+        logger.error("Invalid redirect url: %1", location);
+        return false;
+    }
+
+    // If the url parsing was successful, update the connection parameters
+    m_protocol_envelope = protocol;
+    m_address = address;
+    m_port = port;
+    m_http_request_path_prefix = path;
+    return true;
+}
+
+
 void Connection::websocket_handshake_error_handler(std::error_code ec, const std::string_view* body)
 {
     bool is_fatal;
-    if (ec == util::websocket::Error::bad_response_3xx_redirection ||
-        ec == util::websocket::Error::bad_response_301_moved_permanently ||
-        ec == util::websocket::Error::bad_response_401_unauthorized ||
-        ec == util::websocket::Error::bad_response_5xx_server_error ||
-        ec == util::websocket::Error::bad_response_500_internal_server_error ||
-        ec == util::websocket::Error::bad_response_502_bad_gateway ||
-        ec == util::websocket::Error::bad_response_503_service_unavailable ||
-        ec == util::websocket::Error::bad_response_504_gateway_timeout) {
-        is_fatal = false;
+    logger.error("Handshake error: %1", ec.message());
```

Comment:
Can this mention that the handshake error is a websocket handshake error? Does this basically only happen when we were expecting an HTTP 101 response? Also, do we want to log an error if this is a 301 response that we don't consider an error?